Network tests

akrzos opened the pull request:
4th pass at refactoring of the network tests.
* Pod 2 Pod is working
* Pod 2 Pod with HostNetwork is working with privileged pods
* Svc 2 Svc is working with privileged pods

akrzos commented:
DNM (Do Not Merge) yet as I am still working on the Svc2Svc implementation, however please provide some early feedback.  @jmencak  @sjug  @chaitanyaenr

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+      echo "" > /var/lib/pbench-agent/tools-default/mpstat
+      echo "" > /var/lib/pbench-agent/tools-default/perf
+      echo "" > /var/lib/pbench-agent/tools-default/pidstat
+      master_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/master= | grep master | awk '{print $1}'`
```
`awk '/master/ {print $1}'`
suggest dropping grep, awk can do this for you.  Saying that, all the remaining tests have this issue.

akrzos replied:
Good idea, this actually solves the set -o pipefail issue with clusters without infra nodes.  Would you like to fix the other tests as well?  (Maybe @chaitanyaenr  would be able to post the fix today if available)

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+      for node in $master_nodes; do
+        echo "master" > /var/lib/pbench-agent/tools-default/remote@$node
+      done
+      infra_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/infra= | grep infra | awk '{print $1}'`
```
`awk '/infra/ {print $1}'`

akrzos replied:
See above.  Thanks!

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+      for node in $infra_nodes; do
+        echo "infra" > /var/lib/pbench-agent/tools-default/remote@$node
+      done
+      worker_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/worker= | grep worker | awk '{print $1}'`
```
`awk '/worker/ {print $1}'`

akrzos replied:
ditto!

jmencak commented on workloads/vars/network.yml:
```diff
@@ -0,0 +1,49 @@
+job_completion_poll_attempts: "{{ lookup('env', 'JOB_COMPLETION_POLL_ATTEMPTS')|default(360, true)|int }}"
+
+# Network benchmarks specific parameters:
+network_test_uperf_image: "{{ lookup('env', 'NETWORK_TEST_UPERF_IMAGE')|default('quay.io/openshift-scale/scale-ci-uperf', true) }}"
```
This shuffling of environment variables from to ansible variables and back from ansible variables to environment variables through config maps is very error prone in my opinion, on the other hand, at least the defaults are centralized here.  No solution is probably ideal here.

akrzos replied:
I agree that the shuffling will be a bit confusing at first to new comers but my opinion is the centralizing of the vars is fairly important and helpful in finding what can be changed or adjusted.  I guess it will be a trade-off we might only be able to evaluate over time.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+          fi
+          retries=$[$retries + 1]
+        done
+        [ "${server_ready}" == "false" ] && echo "$(date -u) uperf-server-${rc_num} failed to be ready on time" && exit 1
```
Same as above.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+          fi
+          retries=$[$retries + 1]
+        done
+        [ "${client_ready}" == "false" ] && echo "$(date -u) uperf-client-${rc_num} failed to be ready on time" && exit 1
```
I know that `echo` will probably always exit 0, but is this guaranteed and will this always be the case? If you insist on having one-liners, I suggest:
```
[ "${client_ready}" == "false" ] && { echo "$(date -u) uperf-client-${rc_num} failed to be ready on time"; exit 1; }
```
or more readable variant:
```
[ "${client_ready}" == "false" ] && {
  echo "$(date -u) uperf-client-${rc_num} failed to be ready on time"
  exit 1
}
```
or even better something like a `die()` function?

akrzos replied:
Thanks I will look at the die function and/or implement the more readable function.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+        clients=`oc get pods --output=json -n network0 --selector="name=client" | jq -r '[.items[].status.podIP] | join(",")'`
+        servers=`oc get pods --output=json -n network0 --selector="name=server" | jq -r '[.items[].status.podIP] | join(",")'`
+      fi
+      echo "${clients} ${servers}"
```
Can we have better logging? The uninitiated will probably not know which is which.

akrzos replied:
Ah leftover debugging, I will remove this.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+        sleep 1
+        retries=$[$retries + 1]
+      done
+      [ ${remaining_projects} -gt 0 ] && echo "$(date -u) Failed to clean up network0 project" && exit 1
```
One-liner with a potential issue as above.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+      [ ${remaining_projects} -gt 0 ] && echo "$(date -u) Failed to clean up network0 project" && exit 1
+
+      pbench-copy-results --prefix ${NETWORK_TEST_PREFIX}
+      echo "$(date -u) Completed Network benchmarks with ${pair_count} pairs"
```
There seems to be too many of the `echo "$(date -u)"` commands, use a "log()" function to stderr for more consistency?

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+      kind: ReplicationController
+      metadata:
+        labels:
+          name: "\${NS_VALUE}"
```
I think we should avoid doing things like this.  When you start escaping `$`s in templates, something is wrong, probably the overall design.

akrzos replied:
Good point, I will simplify this template such that it is only an OpenShift template rather than combined env var applied template and OpenShift template.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,961 @@
+        ports:
+        - name: ssh
+          port: ${NETWORK_TEST_UPERF_SSHD_PORT}
+        - name: tcp-20000
```
Cannot we use a range of some sort?  If not, can this be templated?  So many ports copied, there must be another way.

akrzos replied:
I wish there was a range but I couldn't find any implementation of a port range but I did find this open kubernetes issue - https://github.com/kubernetes/kubernetes/issues/23864

Each pair of uperf clients needs one port per uperf "instance" thus the first pair can use 20010-20019, 2nd pair 20020-20029, ...

It actually gets worse, uperf needs a control port opened somewhere between 35000-45000(?) which is problematic for kubernetes services since each port needs to be spelled out.  I am still in the process of understanding how this control port was introduced into pbench-uperf and if we can add a mechanism to control which port it uses so we can precisely control it here.

In the meantime I'll see about moving the ports to be programmatically added.

jmencak replied:
Wouldn't jinja2 for loop be of any help?

jmencak commented on workloads/network.yml:
```diff
@@ -0,0 +1,147 @@
+
+    - name: Label worker nodes when there are no pbench agents for network tests
+      shell: |
+        oc get nodes -l node-role.kubernetes.io/worker= | grep worker | head -n 1 | awk '{print $1}' | xargs -I % oc label nodes % uperf=client
```
Not important, feel free to ignore, but can be simplified with awk doing the grep.  Also suggest using `--no-headers` , but other tests have the same issue.

akrzos replied:
Good point, I will fix this playbook and feel free to fix the others otherwise I will put together a pr to do so.

jmencak replied:
I'll create a PR do this.

akrzos commented:
@jmencak I think I addressed most of your concerns.  I was also able to get the svc2svc benchmarks working completely and if you plan to run a short test with those, use the following uperf container image `quay.io/akrzos/scale-ci-uperf`

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,426 @@
+      echo "" > /var/lib/pbench-agent/tools-default/mpstat
+      echo "" > /var/lib/pbench-agent/tools-default/perf
+      echo "" > /var/lib/pbench-agent/tools-default/pidstat
+      master_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/master= | awk '/master/ {print $1}'`
```
I know I said we can drop the `grep` and move the functionality to `awk`, but is the '/master/' needed now you have a label selector ''node-role.kubernetes.io/master"?  Also, do you need `--show-labels`?

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,426 @@
+      for node in $master_nodes; do
+        echo "master" > /var/lib/pbench-agent/tools-default/remote@$node
+      done
+      infra_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/infra= | awk '/infra/ {print $1}'`
```
Same as above for master.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,426 @@
+      for node in $infra_nodes; do
+        echo "infra" > /var/lib/pbench-agent/tools-default/remote@$node
+      done
+      worker_nodes=`oc get nodes --show-labels -l pbench_agent=true,node-role.kubernetes.io/worker= | awk '/worker/ {print $1}'`
```
Same as above for master.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,426 @@
+          echo -e "    - name: udp-${i}\n      port: ${i}\n      protocol: UDP" >> /tmp/svc-template.yaml
+        done
+      fi
+      eval "echo \"${network_yaml_template}\"" >> /tmp/network.yaml
```
I'm still very uneasy about using eval for this.  It is a powerful friend, but a double-edged sword too.  In this case, you'll lose double quotes, '\\\\' will be evaluated as '\\', there is a possibility of running shell commands through the template...  There might be other side-effects I'm unaware of too.  If this is intended and you know what you're doing, fine, but personally I'd use eval 1) only for environment dumps for reference what variables were used; 2) in cases where I know exactly what I'm doing and have a limited set of input for eval. @sjug, @chaitanyaenr feel free to chime in.

akrzos replied:
I adjusted from `eval` to `envsubst` however we will need https://github.com/openshift-scale/images/pull/10 so the official workload container will have the command.  In the meantime you can use the one built in my namespace to test. (`quay.io/akrzos/scale-ci-workload`)

jmencak replied:
No doubt `envsubst` is safer than `eval`, deferring to other reviewers if they think this is an ok design.

akrzos commented:
@jmencak @sjug @chaitanyaenr PTAL again.  I think I have addressed the major concerns.  If anyone has an advice on making anything look cleaner please speak up.  I will rerun some more versions of the test tonight to verify however I have been able to get all three major scenarios up and running now here (N2N aka P2P w/ HN, P2P, and S2S)

akrzos commented:
I have a few more changes coming for unprivileged workload container performing the tests as well.

akrzos commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,500 @@
+        servers=`oc get pods --output=json -n network0 --selector="name=server" | jq -r '[.items[].status.podIP] | join(",")'`
+      fi
+
+      mv ~/.ssh/config ~/.ssh/config.bak || true
```
@jmencak Let me know if you think we should change this usage of `|| true` it allows privileged and unprivileged execution of this test.  I figured you would be best to review it.

jmencak replied:
Yeah, the joys of `set -e`.  I'd prefer to have a home directory always writable so that this doesn't fail and you don't need `|| true` or similar.  That's why I'm changing `HOME` for my tests.  Didn't run this test yet, so not sure if that's the same issue I had before I changed my `HOME`.

akrzos replied:
Yeah in this case /.ssh/config does not exist and won't exist as we do not put it into the container image.  Even if we did it would be unclear what user should own it as it is the 'unprivileged' execution that hits it. Privileged execution has a $HOME of `/root` and thus find `/root/.ssh/config`.  My opinion is this is that this is an acceptable "1-line" solution.  Another option would be just writing an if statement however that would add a few more lines to each of the two statements that need `|| true`  I like to think that this is an acceptable hack considering it is not complex and that it does the job as needed with less lines.

akrzos commented:
@Deepthidharwar @chaitanyaenr @sjug If you all can take another look, I believe we are close to a final product at this point.  The workload pod can not be unprivileged as well as uperf pods for p2p and s2s, only the hostnetwork pods require privileged execution due to hostnetwork itself.

Let me know your thoughts so we can get this merged in.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,500 @@
+          exit 1
+        fi
+        retries=0
+        while [ $retries -le 60 ] ; do
```
Perhaps move into a function?  Code duplication of the entire client section.

akrzos replied:
Got it, moved this and the log function into a separate functions.sh file sourced at the top.

jmencak commented on README.md:
```diff
@@ -337,3 +337,128 @@ NAMESPACE_CLEANUP
+NETWORK_TEST_UPERF_SSHD_PORT=20000
+NETWORK_TEST_UPERF_PORT_RANGE=""
+NETWORK_TEST_PREFIX=network
+NETWORK_TEST_PAIRS=1
```
I know most of the other tests don't have this, but I don't understand what `NETWORK_TEST_PAIRS` is unless I look through the code.  Documentation of specific test variables would be great not only for this test, but all tests we have. In fact, documenting these vars should be compulsory.

akrzos replied:
In this case pairs refers to client/server uperf pairs. (It gets confusing in that instances refers to how many uperf threads run).  Also I agree in the same sense I don't understand all tests's vars and what are the values for each.  I think this should play into issue #16 though.  The most immediate solution that comes to mind is a docs folder with index and page for each workload but I would like to get some ideas from the group before paving that road.

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,500 @@
+        oc delete project network0
+        retries=0
+        while [ $retries -le 60 ] ; do
+          remaining_projects=`oc get projects | egrep "^network[0-9]+ " -c || true`
```
`--no-headers` would make things cleaner, but this is a nit.

akrzos replied:
NP thanks!

jmencak commented on workloads/files/workload-network-script-cm.yml:
```diff
@@ -0,0 +1,500 @@
+        log "Clean up skipped"
+      fi
+
+      pbench-copy-results --prefix ${NETWORK_TEST_PREFIX}
```
Is it guaranteed for the `NETWORK_TEST_PREFIX` not to contain spaces and special characters? Perhaps using `"` might help at least for folks trying to use spaces in the prefix.

akrzos replied:
Thanks, added quotes.

jmencak commented:
I've tested the `P2P` scenario and it worked out of the box.  Overall, it was a positive experience, even thought the playbook itself actually failed and I didn't let the test itself finish.  Timeout waiting for job completion after 111m57.127s -- perhaps bump up default `job_completion_poll_attempts` for suggested configs.  The test itself was still running after 3h51m having collected 19GB of data.  I've noticed quite a few `*-fail<N>` pbench_uperf directories which started to appear after the first `udp_rr-` test.  I've wiped the cluster after seeing `24-udp_rr-16384B-2i-fail6` which was presumably supposed to be the last iteration.  I love the "Suggested XYZ Test Configuration" in the readme, it should be compulsory for *every* workload, especially more complex ones.  A typical duration of tests running with a suggested configuration would help users too, as this was the first time for me running the network tests and I didn't know what to expect and whether they were running correctly.  Logs and rsh into the `scale-ci-network-xyz` container helped though.  Every long running test should have a `<TEST_NAME>_SMOKE_TEST` option to establish basic test functionality.  You don't want to find out after having run your test for 3 hours you haven't set your `PBENCH_SERVER` or similar properly.  I also believe documentation for workloads show probably tell the users which namespaces the workload creates.  No workload has sufficient documentation at this point.  One last thing I'd recommend is to turn `network0` namespace into a variable and adjust the associated code.
